sudoku_solver/draft_one/lib/sudoku/puzzle.rb

```diff
@@ -0,0 +1,269 @@
+require 'matrix'
+
+require_relative '../n_tree'
+require_relative '../doubly_linked_list'
+
+module Sudoku
+
+  class Puzzle
+    include NTree
+    include DoublyLinkedList
+
+    @@ROWS = [(0..8),
+              (9..17),
+              (18..26),
+              (27..35),
+              (36..44),
+              (45..53),
+              (54..62),
+              (63..71),
+              (72..80)]
+
+    attr_accessor :parent, :children, :next, :prev
+    attr_reader :head, :tail, :last_sibling
+    # @param [Array<Integer>] blocks
+    # @param [Puzzle] parent
+    # @param [Puzzle] children
+    # @param [Puzzle] nxt
+    # @param [Puzzle] prev
+    # @param [Puzzle] head
+    # @param [Puzzle] tail
+    def initialize(blocks, parent = nil, children = nil, nxt = nil, prev = nil, head = nil, tail = nil)
```

Comment:
So, I'll be honest, I don't really understand the purpose of the `DoublyLinkedList` in this class.  There's a fancy way to solve a Sudoku puzzle with a famous algorithm called [Dancing Links](http://en.wikipedia.org/wiki/Dancing_Links) which is typically implemented using doubly linked lists, but that's not what you're doing here.

For the tree-centric solution, make sure you're clear on what tree we're talking about.  I think you are, but I'll restate just in case.

First, you have to imagine a big tree where each node is a Sudoku board.  Don't worry for now about what each Sudoku board looks like; just make sure the idea of a tree where the values are Sudoku boards makes sense.

Next, let's talk about what the tree actually looks like.  For any tree, the key thing to understand is this: "given a node, what are its children?"  If we have an arbitrary Sudoku board — maybe finished, maybe partially finished, maybe started — is there some sensible and useful way we can define the "children" of _that_ specific board?

Yes, and here's how.

Start with an incomplete Sudoku board and single out one and only one blank square to designate as the "guessing square" for the board.  For the sake of simplicity, let's say we always pick the first blank square as the guessing square .   By looking at the numbers in the squares in the same row/column/box as the guessing square, we'll get anywhere between 0 and 9 potential numbers that could go in the guessing square.

- If there are 0 potential numbers, this board has no children (and we also know it's impossible to solve)
- Otherwise, let this board have one "child board" for every number that could potentially go in the guessing square. Each child board is identical to the original board except with the original board's guessing square filled in with one of the potential numbers.

In other words, if the first blank square of a Sudoku board could potentially be filled with `1,6,8` then that board has three child boards.  The first child board is identical to the original except we've filled in the original board's "guessing square" with a `1`.  The second and third children have that blank space filled in with `6` and `8`, respectively.

Each child board has its own guessing square, too.

This allows us to build a tree from the top down, with the original starting board as the root.  A board has no children in one of two scenarios: there are no blank squares with potential guesses or every square has been filled in.  By definition, nodes in a tree without children are leaf nodes.

So, if we do a depth-first search of this tree, we'll eventually find a solution in one of the leaf nodes.